src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Syntax/SyntaxTokenList.cs

```diff
@@ -0,0 +1,497 @@
+    }
+
+    public SyntaxTokenList Remove(SyntaxToken tokenInList)
+    {
+        var index = IndexOf(tokenInList);
+        return index >= 0 ? RemoveAt(index) : this;
+    }
+
+    public SyntaxTokenList Replace(SyntaxToken tokenInList, SyntaxToken newToken)
+    {
+        if (newToken == default)
+        {
+            ThrowHelper.ThrowArgumentOutOfRangeException(nameof(newToken));
+        }
+
+        return ReplaceRange(tokenInList, [newToken]);
+    }
+
+    public SyntaxTokenList ReplaceRange(SyntaxToken tokenInList, ReadOnlySpan<SyntaxToken> tokens)
+    {
+        var index = IndexOf(tokenInList);
+
+        if (tokens.Length == 0)
+        {
+            return RemoveAt(index);
+        }
+
+        if (index < 0)
+        {
+            ThrowHelper.ThrowArgumentOutOfRangeException(nameof(tokenInList));
+        }
+
+        var count = Count;
+        var array = new ArrayElement<GreenNode>[count + tokens.Length - 1];
+
+        // Add current tokens up to 'index'
+        int i;
+        for (i = 0; i < index; i++)
+        {
+            array[i].Value = this[i].RequiredNode;
+        }
+
+        // Add new tokens
+        for (var j = 0; j < tokens.Length; i++, j++)
+        {
+            array[i].Value = tokens[j].RequiredNode;
+        }
+
+        // Add remaining tokens starting *after* 'index'
+        for (var j = index + 1; j < count; i++, j++)
+        {
+            array[i].Value = this[j].RequiredNode;
+        }
+
+        return new(parent: null, InternalSyntax.SyntaxList.List(array), position: 0, index: 0);
+    }
+
+    public SyntaxTokenList ReplaceRange(SyntaxToken tokenInList, IEnumerable<SyntaxToken> tokens)
+    {
+        var index = IndexOf(tokenInList);
+
+        if (index < 0)
+        {
+            ThrowHelper.ThrowArgumentOutOfRangeException(nameof(tokenInList));
+        }
+
+        ArgHelper.ThrowIfNull(tokens);
+
+        if (tokens.TryGetCount(out var tokenCount))
+        {
+            return ReplaceRangeWithCount(index, tokens, tokenCount);
+        }
+
+        var count = Count;
+        using var builder = new PooledArrayBuilder<SyntaxToken>(count);
+
+        // Add current tokens up to 'index'
+        for (var i = 0; i < index; i++)
+        {
+            builder.Add(this[i]);
+        }
+
+        // Add new tokens
+        foreach (var token in tokens)
+        {
+            builder.Add(token);
+        }
+
+        // Add remaining tokens starting *after* 'index'
+        for (var i = index + 1; i < count; i++)
+        {
+            builder.Add(this[i]);
+        }
+
+        return new(parent: null, builder.ToGreenListNode(), position: 0, index: 0);
+    }
+
+    private SyntaxTokenList ReplaceRangeWithCount(int index, IEnumerable<SyntaxToken> tokens, int tokenCount)
+    {
+        if (tokenCount == 0)
+        {
+            return RemoveAt(index);
+        }
+
+        var count = Count;
+        var array = new ArrayElement<GreenNode>[count + tokenCount];
```

Comment:
Same question aboug `tokenCount` usage.

Reply:
Will add a `Debug.Assert(...)` that verifies `i == index + tokenCount` and unit tests that cover it.